### PR TITLE
Refactor ADM classes

### PR DESCRIPTION
**Pull request by ebsmothers**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #490
* #489
* __->__ #488

Differential Revision: [D50288361](https://our.internmc.facebook.com/intern/diff/D50288361)

### CODE REVIEW

**Comment by ebsmothers** [comment]
@ebsmothers has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50288361).

**Comment by ebsmothers** [comment]
@ebsmothers has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50288361).

File: torchmultimodal/diffusion_labs/models/adm_unet/adm.py [hunk 1]
```diff
@@ -24,25 +24,55 @@
     SinusoidalPositionEmbeddings,
 )
 
+DEFAULT_EMBED_NAME = "clip_image"
```
Comment (pbontrager) [inline torchmultimodal/diffusion_labs/models/adm_unet/adm.py:27]: I think this should be a parameter for ADMUnet instead of a global variable. If you look at the transform and adapter classes they all take the conditional input key as "*_field", for this case it could be "conditional_field". Also I don't think the default name should be based on CLIP as that's a dalle input and not generic to ADM. It should be something generic like "context" or "condition".

File: torchmultimodal/diffusion_labs/models/adm_unet/adm.py [hunk 2]
```diff
@@ -68,132 +98,6 @@ class ADM(nn.Module):
...
-
-    Args:
-        x (Tensor): input Tensor of shape [b, in_channels, h, w]
-        res_conditional_embedding (Tensor): conditioning embedding vector of shape [b, dim_res_cond].
-        attn_conditional_embedding (Tensor): inputs to be used in cross attention of shape [b, n, dim_attn_cond] where n is the
-            number of tokens.
-    """
-
     def __init__(
         self,
```
Comment (pbontrager) [inline torchmultimodal/diffusion_labs/models/adm_unet/adm.py:102]: We should add "*" for forced named parameters

File: torchmultimodal/diffusion_labs/models/adm_unet/adm.py [hunk 3]
```diff
@@ -378,7 +346,65 @@ def forward(
             if hidden_states:
                 h = torch.cat([h, hidden_states.pop()], dim=1)
             h = block(h, res_conditional_embedding, attn_conditional_embedding)
-        return h
+
+        # If model is predicting variance, then it should be configured to output double the channels as input
```
Comment (pbontrager) [inline torchmultimodal/diffusion_labs/models/adm_unet/adm.py:350]: This should probably be broken out as it's own method (_get_variance_value) to clean up the forward method a bit. If predict_variance_value is false it can return None. Then you can just return DiffusionOutput with both prediction and variance_value as the default variance_value is None anyway.

File: torchmultimodal/diffusion_labs/models/adm_unet/adm.py [hunk 4]
```diff
@@ -378,7 +346,65 @@ def forward(
...
+            return DiffusionOutput(
+                prediction=prediction,
+                variance_value=variance_value,
+            )
+        else:
+            return DiffusionOutput(
+                prediction=h,
+            )
+
+    def _get_conditional_projections(
```
Comment (pbontrager) [inline torchmultimodal/diffusion_labs/models/adm_unet/adm.py:369]: nit I think it's better to have forward as the last method in a nn.Module so it's easier to find.

File: torchmultimodal/diffusion_labs/models/adm_unet/adm.py [hunk 5]
```diff
@@ -453,7 +479,7 @@ def adm_unet(
     time_embed_dim: int = 512,
```
Comment (pbontrager) [inline torchmultimodal/diffusion_labs/models/adm_unet/adm.py:479]: Do all of these parameters make sense here anymore? Since a user can just call ADMUnet if they want to change many defaults. Maybe just expose "embedding_name", "image_channels", and "predict_variance"

File: torchmultimodal/diffusion_labs/models/adm_unet/adm.py [hunk 6]
```diff
@@ -453,7 +479,7 @@ def adm_unet(
...
     cond_embed_dim: int = 2048,
     clip_embed_dim: int = 768,
```
Comment (pbontrager) [inline torchmultimodal/diffusion_labs/models/adm_unet/adm.py:481]: These two variables should be just "embed_*" without clip

File: torchmultimodal/diffusion_labs/models/adm_unet/adm.py [hunk 7]
```diff
@@ -453,7 +479,7 @@ def adm_unet(
...
-    clip_embed_name: str = "clip_image",
+    clip_embed_name: str = DEFAULT_EMBED_NAME,
     predict_variance_value: bool = True,
     # ADMUNet args
     image_channels: int = 4,
```
Comment (pbontrager) [inline torchmultimodal/diffusion_labs/models/adm_unet/adm.py:485]: Can't comment lower than this line. But on line 490 I think this should be: constructs a conditional ADM U-Net

File: torchmultimodal/diffusion_labs/models/adm_unet/adm.py [hunk 8]
```diff
@@ -202,10 +106,40 @@ def __init__(
         use_attention_for_layer: List[bool],
         dim_res_cond: int,
         dim_attn_cond: Optional[int] = None,
+        embed_dim: Optional[int] = None,
         in_channels: int = 3,
         out_channels: int = 3,
+        time_embed_dim: Optional[int] = None,
+        timestep_encoder: Optional[nn.Module] = None,
```
Comment (pbontrager) [inline torchmultimodal/diffusion_labs/models/adm_unet/adm.py:113]: I'm not sure if it makes sense to make these three encoders/projectors a parameter. It makes the init much more messy. I'd be in favor of removing that option.

**Review by pbontrager — COMMENTED** [review]
Looks great overall and structurally good. I left some comments around name and parameters.

File: torchmultimodal/diffusion_labs/models/adm_unet/adm.py (hunk 5 above)
Comment (pbontrager) [reply torchmultimodal/diffusion_labs/models/adm_unet/adm.py:479]: I just mean, that if the purpose of this builder is just to house the defaults used in ADM then I don't think we should allow the user to override all of the defaults. If they need custom parameters they can use ADMUnet. If they want the ADM defaults then they can use this functions with just the three parameters exposed I mentioned. I'm also okay with moving these defaults to ADMUnet and dropping this builder.

File: torchmultimodal/diffusion_labs/models/adm_unet/adm.py (hunk 8 above)
Comment (ebsmothers) [reply torchmultimodal/diffusion_labs/models/adm_unet/adm.py:115]: After our offline discussion, I plan to leave these in for this version to unblock other changes. But let's remove in a follow-up

File: torchmultimodal/diffusion_labs/models/adm_unet/adm.py (hunk 5 above)
Comment (ebsmothers) [reply torchmultimodal/diffusion_labs/models/adm_unet/adm.py:489]: I see. Maybe let's do that in a follow-up then? I think dalle2_decoder is using a bunch of these params rn so making this change would have implications for that

File: torchmultimodal/diffusion_labs/models/adm_unet/adm.py (hunk 1 above)
Comment (ebsmothers) [reply torchmultimodal/diffusion_labs/models/adm_unet/adm.py]: Updated to context, left clip_image in the builder. Lmk if that's what you have in mind

**Comment by ebsmothers** [comment]
@ebsmothers has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50288361).

File: torchmultimodal/diffusion_labs/models/adm_unet/adm.py (hunk 1 above)
Comment (pbontrager) [reply torchmultimodal/diffusion_labs/models/adm_unet/adm.py]: CLIP is a dalle2 input and is not specific to ADM. So I think we should remove the name from everywhere. Also I still don't believe this should be a global variable.

File: torchmultimodal/diffusion_labs/models/adm_unet/adm.py (hunk 5 above)
Comment (pbontrager) [reply torchmultimodal/diffusion_labs/models/adm_unet/adm.py:489]: These defaults match dalle2 decoder. But you're right we should update dalle2 decoder then to be simpler.

File: torchmultimodal/diffusion_labs/models/adm_unet/adm.py [hunk 9]
```diff
@@ -471,8 +507,8 @@ def adm_unet(
     Args:
         time_embed_dim (int): desired dimensionality of timestep embedding
         cond_embed_dim (int): desired dimensionality of conditional input embeddings
-        clip_embed_dim (int): expected dimensionality of CLIP image embeddings
-        clip_embed_name (str): name of CLIP embedding conditional input
+        embed_dim (int): expected dimensionality of CLIP image embeddings
```
Comment (pbontrager) [inline torchmultimodal/diffusion_labs/models/adm_unet/adm.py:510]: The description here shouldn't say CLIP either. It's the expected dim of the conditional image embedding

**Comment by ebsmothers** [comment]
@ebsmothers has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50288361).

File: torchmultimodal/diffusion_labs/models/adm_unet/adm.py (hunk 1 above)
Comment (ebsmothers) [reply torchmultimodal/diffusion_labs/models/adm_unet/adm.py]: Ditched the global variable and used "context" for default in both the ADMUNet class and the builder

**Review by pbontrager — APPROVED** [review]
It looks good now. Thanks for merging these two classes!
